File: src/main/java/at/searles/buf/FrameStreamImpl.java
package at.searles.buf;

import org.jetbrains.annotations.NotNull;

public class FrameStreamImpl implements FrameStream {

    private final BufferedStream.Impl stream;
    private final CharSeq frame;
    private long frameStart;
    private long frameEnd;
    private boolean invalid;

    public FrameStreamImpl(Impl stream) {
        this.stream = stream;
        this.invalid = false;
        this.frame = new CharSeq();
    }

    @Override
    public int next() {
        // is the frame too big?
        if (frameEnd - frameStart >= stream.bufSize()) {
            throw new IllegalArgumentException("buffer size is too small");
        }

        return stream.next();
    }

    @Override
    public long position() {
        return stream.position();
    }

    @Override
    public void setPositionTo(long ptr) {
        invalid = true;
        stream.setPositionTo(ptr);
        this.frameStart = ptr;
        this.frameEnd = ptr;
    }

    @Override
    public void mark() {
        invalid = true;
        this.frameEnd = stream.position();
    }

    @Override
    public void advance() {
        invalid = true;
        this.frameStart = this.frameEnd;
        stream.setPositionTo(this.frameEnd);
    }

    @Override
    public void reset() {
        invalid = true;
        this.frameEnd = this.frameStart;
        stream.setPositionTo(this.frameStart);
    }

    @Override
    public Frame frame() {
        return frame;
    }

    private class CharSeq implements Frame {
        private String str;

        private void update() {
            if (invalid) {
                StringBuilder sb = new StringBuilder();

                for (long index = frameStart; index < frameEnd; ++index) {
                    sb.appendCodePoint(stream.cached(index));
                }

                str = sb.toString();
                invalid = false;
            }
        }

        @Override
        public long startPosition() {
            return frameStart;
        }

        @Override
        public long endPosition() {
            return frameEnd;
        }

        @Override
        public int length() {
            update();
            return str.length();
        }

        @Override
        public char charAt(int index) {
            update();
            return str.charAt(index);
        }

        @Override
        public CharSequence subSequence(int start, int end) {
            update();
            return str.subSequence(start, end);
        }

        @NotNull
        public String toString() {
            update();
            return str;
        }
    }
}

File: src/main/java/at/searles/buf/FrameStreamImpl.kt
package at.searles.buf

class FrameStreamImpl(private val stream: BufferedStream.Impl) : FrameStream {
    override val frame: Frame = CharSeq()
    private var frameStart: Long = 0
    private var frameEnd: Long = 0
    private var invalid = false

    override fun next(): Int {
        // is the frame too big?
        require(frameEnd - frameStart < stream.bufSize()) { "buffer size is too small" }
        return stream.next()
    }

    override var position: Long
        get() {
            return stream.position
        }
        set(value) {
            invalid = true
            stream.position = value
            frameStart = value
            frameEnd = value
        }

    override fun mark() {
        invalid = true
        frameEnd = stream.position
    }

    override fun advance() {
        invalid = true
        frameStart = frameEnd
        stream.position = frameEnd
    }

    override fun reset() {
        invalid = true
        frameEnd = frameStart
        stream.position = frameStart
    }

    private inner class CharSeq : Frame {
        private lateinit var str: String

        private fun update() {
            if (invalid) {
                val sb = StringBuilder()

                for (index in frameStart until frameEnd) {
                    sb.appendCodePoint(stream.cached(index))
                }

                str = sb.toString()
                invalid = false
            }
        }

        override val start: Long get() = frameStart

        override val end: Long get() = frameEnd
        override val length: Int get() {
            update()
            return str.length
        }

        override fun get(index: Int): Char {
            update()
            return str[index]
        }

        override fun subSequence(startIndex: Int, endIndex: Int): CharSequence {
            update()
            return str.subSequence(startIndex, endIndex)
        }

        override fun toString(): String {
            update()
            return str
        }
    }
}